pipeline/metadata/add_metadata.py

```diff
@@ -75,17 +109,96 @@ def annotate_row_ip(
 
     return rows_with_metadata
 
+  def annotate_answer_ips(
+      self, rows: beam.pvalue.PCollection[SatelliteRow]
+  ) -> beam.pvalue.PCollection[SatelliteRow]:
+    """Add ip metadata info to received ip lists in rows
+
+    Args:
+      rows: PCollection of SatelliteRows
+
+    Returns:
+      SatelliteRows with routeview metadata attatched to answer ips.
+    """
+    # PCollection[Tuple[roundtrip_id, SatelliteRow]]
+    rows_with_roundtrip_id = (
+        rows | 'add roundtrip_ids: answer metadata' >>
+        beam.Map(set_random_roundtrip_id).with_output_types(
+            Tuple[str, SatelliteRow]))
+
+    # PCollection[Tuple[DateIpKey, Tuple[roundtrip_id, SatelliteAnswer]]]
+    received_ips_keyed_by_ip_and_date = (
+        rows_with_roundtrip_id |
+        'get received ips: answer metadata' >> beam.FlatMap(
+            _get_received_ips_with_roundtrip_id_and_date).with_output_types(
+                Tuple[DateIpKey, Tuple[str, SatelliteAnswer]]))
+
+    #PCollection[DateIpKey]
+    # pylint: disable=no-value-for-parameter
+    ips_and_dates = (
```

Comment:
re: this and the function/transform comment below. I agree that would be cleaner, but currently this function has the same structure as the other two triple cogroups [add_received_ip_tags](https://github.com/censoredplanet/censoredplanet-analysis/blob/b007d8f5eb3ab109026bcc0ad388849e0fd7c3fb/pipeline/metadata/satellite.py#L455) and [add_page_fetch_to_answers](https://github.com/censoredplanet/censoredplanet-analysis/blob/b007d8f5eb3ab109026bcc0ad388849e0fd7c3fb/pipeline/metadata/satellite.py#L586).

Refactoring the three of them together will involve adding some of the generic beam transforms you're discussing to be reused between them. But it will be easier to do if we keep the parallel structure for now.